Potential Bug around geo coordinate properties clean up

When deleting a class (with data imported) that has geo coordinates it seems clean up is not working properly. During an unrelated test with a 6-node cluster, we could see the following symptoms:

* There is still geo-coordinates-related state on disk
* It seems the clean-up processes keep running even when the backing DB has been deleted, this led to a nil-pointer panic that could even crash the entire machine.

Further Analysis:

The issue occurs when there is an async clean-up to be done (e.g. after a delete/update), but the class was deleted before the async process runs. Instead of the deletion stopping the process, it continues. This makes no sense, since the LSM store is already gone, as we can see with the following panic:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x70 pc=0x18ca977]

goroutine 449416 [running]:
github.com/semi-technologies/weaviate/adapters/repos/db/lsmkv.(*segment).getBySecondary(0x0, 0x0, 0xc003cd4680, 0x8, 0x40, 0xc001b19740, 0x18c2159, 0xc00dbdc880, 0xc002916f60, 0xc003cd4680)
```

This issue is specific only to `geoCoordinates`.

To reproduce you can do the following:

1. Import Demo dataset ("NewsPublications") with Geo Coordinates
1. Import the same dataset again without deleting the classes beforehand (This will lead to many updates)
1. Delete all the classes
1. Wait for the next HNSW cleanup cycle to run on the deleted property - it will produce the panic shown above.

--- a/adapters/repos/db/propertyspecific/index.go
+++ b/adapters/repos/db/propertyspecific/index.go
@@ -12,6 +12,7 @@
 package propertyspecific
 
 import (
+	"github.com/pkg/errors"
 	"github.com/semi-technologies/weaviate/adapters/repos/db/vector/geo"
 	"github.com/semi-technologies/weaviate/entities/schema"
 )
@@ -34,4 +35,22 @@ type Indices map[string]Index
 func (i Indices) ByProp(propName string) (Index, bool) {
 	index, ok := i[propName]
 	return index, ok
+}
+
+func (i Indices) DropAll() error {
+	for propName, index := range i {
+		if index.Type != schema.DataTypeGeoCoordinates {
+			return errors.Errorf("no implementation to delete property %s index of type %v",
+				propName, index.Type)
+		}
+
+		if err := index.GeoIndex.Drop(); err != nil {
+			return errors.Wrapf(err, "drop property %s", propName)
+		}
+
+		index.GeoIndex = nil
+		delete(i, propName)
+
+	}
+	return nil
 }

--- a/adapters/repos/db/shard.go
+++ b/adapters/repos/db/shard.go
@@ -166,6 +166,11 @@ func (s *Shard) drop() error {
 	// TODO: can we remove this?
 	s.deletedDocIDs.BulkRemove(s.deletedDocIDs.GetAll())
 
+	err = s.propertyIndices.DropAll()
+	if err != nil {
+		return errors.Wrapf(err, "remove property specific indices at %s", s.DBPathLSM())
+	}
+
 	return nil
 }
 

--- a/adapters/repos/db/vector/geo/geo.go
+++ b/adapters/repos/db/vector/geo/geo.go
@@ -43,6 +43,7 @@ type vectorIndex interface {
 		allowList helpers.AllowList) ([]uint64, error)
 	Delete(id uint64) error
 	Dump(...string)
+	Drop() error
 }
 
 // Config is passed to the GeoIndex when its created
@@ -76,6 +77,15 @@ func NewIndex(config Config) (*Index, error) {
 	}
 
 	return i, nil
+}
+
+func (i *Index) Drop() error {
+	if err := i.vectorIndex.Drop(); err != nil {
+		return err
+	}
+
+	i.vectorIndex = nil
+	return nil
 }
 
 func makeCommitLoggerFromConfig(config Config) hnsw.MakeCommitLogger {

